### PR TITLE
1.0.0 dylandev

### PR DESCRIPTION
**Pull request by dm0896665**
Added in Character/General page with 3D model. Also fixed some styling/ user bugs.

### CODE REVIEW

File: lib/sacastats_web/controllers/character_controller.ex [hunk 1]
```diff
@@ -20,10 +21,14 @@ defmodule SacaStatsWeb.CharacterController do
         |> redirect(to: Routes.character_path(conn, :character_search))
 
       {:ok, %PS2.API.QueryResult{data: body}} ->
+        next_rank = Map.get(body, "battle_rank")
+        {next_rank, ""} = Integer.parse(Map.get(next_rank, "value"))
+        next_rank = next_rank + 1
```
Comment (Bentheburrito) [inline lib/sacastats_web/controllers/character_controller.ex:26]:
```suggestion
        next_rank = 
          body
          |> get_in(["battle_rank", "value"])
          |> String.to_integer()
          |> Kernel.+(1)
```
One feature I love in Elixir is pipes (`|>`). The pipe operator takes the result of the expression on the left side (or in this case, on the line above), and uses it as the first argument in the function call on the right. E.g. `"26" |> String.to_integer()` is the same as `String.to_integer("26")`.

`get_in/2` is also a cool function - it makes it very easy to get nested values.

File: lib/sacastats_web/controllers/character_controller.ex [hunk 2]
```diff
@@ -20,10 +21,14 @@ defmodule SacaStatsWeb.CharacterController do
...
+
         character_stuff = %{
```
Comment (Bentheburrito) [inline lib/sacastats_web/controllers/character_controller.ex:28]:
```suggestion
        character = %{
```
Maybe we can rename `character_stuff` to just `character`.

File: lib/sacastats_web/controllers/character_controller.ex [hunk 3]
```diff
@@ -20,10 +21,14 @@ defmodule SacaStatsWeb.CharacterController do
...
-          "name" => name,
           "stat_page" => String.downcase(stat_template_name) <> ".html",
-          "response" => Map.get(body, "character_id")
+          "response" => body,
+          "next_rank" => next_rank
         }
 
         render(conn, "template.html", character: character_stuff)
```
Comment (Bentheburrito) [inline lib/sacastats_web/controllers/character_controller.ex:34]:
```suggestion
        render(conn, "template.html", character: character)
```
rename as suggested above

File: lib/sacastats_web/templates/character/general.html.heex [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+<div id="characterGeneral" class="row">
+    <div id="characterModelHolder" class="col-md-4 d-none d-md-flex">
+        <figure id="characterModel" class="w-100"></figure>
+        <!-- <img src={Routes.static_path(@conn, "/images/logo_icon.ico")} alt="Sacagawea Icon"> -->
+    </div>
+    <div class="col-md-8 col-12">
+        <div class="stat-section">
+            <h1 id="characterName"><%= @character["response"]["name"]["first"] %></h1>
+            <div class="d-none d-sm-flex float-sm-right">
+                <%= if Integer.parse(@character["response"]["online_status"]) > {0,""} do %>
```
Comment (Bentheburrito) [inline lib/sacastats_web/templates/character/general.html.heex:12]: Could you change this to be similar to what you did with `next_rank` in the controller? I.e. parse `online_status` in the controller via a pipeline and pass it as its own field in the `character` assign.

File: lib/sacastats_web/templates/character/general.html.heex [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    <img src={Routes.static_path(@conn, "/images/character/online.ico")} alt="online" width="30">
+                    <p class="d-inline my-auto">Online</p>
+                <% else %>
+                    <img src={Routes.static_path(@conn, "/images/character/offline.ico")} alt="offline" width="30">
+                    <p class="d-inline my-auto">Offline</p>
+                <% end %>
+            </div>
+            <p id="characterId"><%= @character["response"]["character_id"] %></p>
+            <div class="d-sm-none d-flex justify-content-center pb-3">
+                <%= if Integer.parse(@character["response"]["online_status"]) > {0,""} do %>
```
Comment (Bentheburrito) [inline lib/sacastats_web/templates/character/general.html.heex:22]: See suggestion above ^^

File: lib/sacastats_web/templates/layout/root.html.heex [hunk 6]
```diff
@@ -1,30 +1 @@
-<!DOCTYPE html>
```
Comment (Bentheburrito) [inline lib/sacastats_web/templates/layout/root.html.heex:1]: Is it okay to delete this !DOCTYPE tag? I didn't notice any issues while testing locally but just want to be sure.

File: lib/sacastats_web/templates/page/login.html.heex [hunk 7]
```diff
@@ -2,4 +2,4 @@
     <div class="col-sm-6 col-xl-3 col-xs-12 mx-auto">
       <%= link "Login with Discord", to: "/login/new", class: "btn btn-primary" %>
     </div>
-</div>
+</div>
```
Comment (Bentheburrito) [inline lib/sacastats_web/templates/page/login.html.heex:5]:
```suggestion
</div>

```
Newline - I will look into some GitHub actions/custom commit flow to format automatically for us

File: priv/static/css/character/general.css [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+.move-up-4{
+    position: relative;
+    top: -4rem !important;
+}
+@media only screen and (max-width: 767px) { /*for boostrap xs*/
+    .move-up-4{
+        position: relative;
+        top: -2rem !important;
+    }
+}
```
Comment (Bentheburrito) [inline priv/static/css/character/general.css:38]:
```suggestion
}

```

File: priv/static/css/character/lookup.css [hunk 9]
```diff
@@ -0,0 +1,3 @@
+.form-control {
+    font-size: 14;
+}
```
Comment (Bentheburrito) [inline priv/static/css/character/lookup.css:3]:
```suggestion
}

```

File: priv/static/css/character/template.css [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    padding-bottom: 0px;
+}
+.pb-10 {
+    padding-bottom: 10rem !important;
+}
+@media only screen and (max-width: 767px) { /*for boostrap xs*/
+    .main-content{
+      padding-bottom: 18rem !important;
+    }
+}
```
Comment (Bentheburrito) [inline priv/static/css/character/template.css:18]:
```suggestion
}

```

File: priv/static/js/character/lookup.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        swapURL(event);
+    });
+}
+
+function swapURL(event) {
+    showLoadingScreen();
+    event.preventDefault();
+    let character = document.getElementById("character").value;
+    location.href = document.URL + "/" + character;
+}
```
Comment (Bentheburrito) [inline priv/static/js/character/lookup.js:21]:
```suggestion
}

```

File: priv/static/js/character/model.js [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    loadModels();
+    createRenderer();
+
+    renderer.setAnimationLoop(() => {
+        update();
+        render();
+    });
+
+    window.addEventListener("resize", onWindowResize);
+}
```
Comment (Bentheburrito) [inline priv/static/js/character/model.js:137]:
```suggestion
}

```
Very cool - you'll have to show me some Three.JS stuff sometime ;P

**Review by Bentheburrito — CHANGES_REQUESTED** [review]
This is awesome!! It feels so responsive and smooth to navigate between pages. Really nice work. 
I left a few suggestions on code style/end-of-file newlines, otherwise it all looks great.

File: lib/sacastats_web/templates/layout/root.html.heex (hunk 6 above)
Comment (dm0896665) [reply lib/sacastats_web/templates/layout/root.html.heex:1]: Yeah, that was an integrated HTML page, so it was just a duplicate that I removed.

File: lib/sacastats_web/templates/character/general.html.heex (hunk 4 above)
Comment (dm0896665) [reply lib/sacastats_web/templates/character/general.html.heex:12]: So is it okay to have a massive controller class? I know that in Springboot you really want to keep your controller classes small.

File: lib/sacastats_web/controllers/character_controller.ex (hunk 2 above)
Comment (dm0896665) [reply lib/sacastats_web/controllers/character_controller.ex:28]: Good call.

File: lib/sacastats_web/controllers/character_controller.ex (hunk 1 above)
Comment (dm0896665) [reply lib/sacastats_web/controllers/character_controller.ex:26]: Yeah I'm not the best with the elixir piping yet lol

File: lib/sacastats_web/templates/character/general.html.heex (hunk 4 above)
Comment (Bentheburrito) [reply lib/sacastats_web/templates/character/general.html.heex:12]: I think right now the size isn't too bad. Once all of the stat pages are in place, then yeah we might want a separate module of helper classes to make things more concise in the controllers, or even split CharacterController into smaller controllers?

Comment (dm0896665) [reply lib/sacastats_web/templates/character/general.html.heex]: Not sure how to update the changes here, but they should be pushed.

**Review by Bentheburrito — APPROVED** [review]

**Comment by Bentheburrito** [comment]
Thanks! Feel free to delete `1.0.0-dylandev` at your leisure :)

**Pull request merged**